internal/abci/abci.go

```diff
@@ -365,7 +368,15 @@ func (a *AbciApp) CheckTx(ctx context.Context, incoming *abciTypes.RequestCheckT
 	}
 	defer readTx.Rollback(ctx) // always rollback since we are read-only
 
-	err = a.txApp.ApplyMempool(ctx, readTx, tx)
+	err = a.txApp.ApplyMempool(&common.TxContext{
+		Ctx: ctx,
+		BlockContext: &common.BlockContext{
+			ChainContext: a.chainContext,
+			Height:       a.height + 1, // height increments at the start of FinalizeBlock,
+			Proposer:     nil,          // we don't know the proposer here
+		},
+		TxID: cometTXID(incoming.Tx),
```

Comment:
TxID is this used somewhere else? better to avoid this as much as possible, as it has huge performance impact. Also, its recalculated below 403-408 lines. 
```
if newTx {
		txHash := sha256.Sum256(incoming.Tx)
		a.verifiedTxnsMtx.Lock()
		a.verifiedTxns[txHash] = struct{}{}
		a.verifiedTxnsMtx.Unlock()
	}
```

Reply:
Ah good point

Reply:
Wait. this is not used elsewhere

Reply:
It is used directly above, but that is in an error return, so in either case, it is hashed once.